### PR TITLE
Support for new g6 firmware

### PR DESCRIPTION
**Pull request by efidoman**
- fix last entry timing issue that caused some missed readings while offline
- support auto tx restarts by age based upon 7 days for g5 and 10 days for g6
- support new g6 firmware by not exiting if unfiltered is null and glucose value is valid

### CODE REVIEW

File: xdrip-get-entries.sh [hunk 1]
```diff
@@ -1060,25 +1078,28 @@ function  capture_entry_values()
...
+  # check for valid and not expired sessionStartDate
+  sessionMinutesRemaining=$(bc <<< "($sessionMaxSeconds - ($epochdate-$sessionStartDateEpochms/1000))/60")
+  if [ $(bc <<< "$sessionMinutesRemaining < 0") -eq 1 -a $(bc <<< "$sessionMinutesRemaining > ($sessionMaxSeconds * 60)") -eq 1 ]; then
     log "Expired session or invalid sessionStartDate, not processing auto-restart logic"
-    sessionMinutesRemaining=100000 # big number ensures auto-restart logic will not kick in
   fi
   log "sessionStartDate=$sessionStartDate, sessionStartDateEpochms=$sessionStartDateEpochms" 
   log "sessionMinutesRemaining=$sessionMinutesRemaining"
   if [ $(bc <<< "$sessionMinutesRemaining < 65") -eq 1 ]; then
     #TODO use parameter to do this only if auto-restart is true
```
Comment (tynbendad) [inline xdrip-get-entries.sh]: you can get rid of the #TODO here

File: xdrip-get-entries.sh [hunk 2]
```diff
@@ -1060,25 +1078,28 @@ function  capture_entry_values()
...
     if [ $(bc <<< "$glucose < 400") -eq 1  -a $(bc <<< "$glucose > 40") -eq 1 ]; then
       if [ $(bc <<< "$variation < 10") -eq 1 ]; then
+        if [[ "$auto_sensor_restart" == true ]]; then
          cgm-stop; sleep 1; cgm-start -m 120; sleep 1; cgm-calibrate $glucose; sleep 1; cgm-calibrate $glucose
```
Comment (tynbendad) [inline xdrip-get-entries.sh]: this is amazingly cool if it works!

File: xdrip-get-entries.sh [hunk 3]
```diff
@@ -1860,8 +1881,8 @@ function check_last_glucose_time_smart_sleep()
 {
   rm -f $xdripMessageFile
 
-  if [ -e $lastEntryFile ]; then
-    entry_timestamp=$(date -r $lastEntryFile +'%s')
+  if [ -e ${LDIR}/entry-watchdog ]; then
```
Comment (tynbendad) [inline xdrip-get-entries.sh]: oh, i see it now - disregard my review comment about not seeing this!

File: xdrip-get-entries.sh (hunk 1 above)
Comment (efidoman) [reply xdrip-get-entries.sh]: Thanks - will do.

File: xdrip-get-entries.sh (hunk 2 above)
Comment (efidoman) [reply xdrip-get-entries.sh]: It worked for me at least once already. It takes a couple of cycles for the tx to say OK, but it did work.

**Comment by efidoman** [comment]
I'm testing the auto_sensor_restart feature again now. I'm at my second 9 days for this sensor so I put some test code in to make it think it is within 65 minutes of 10 days.

I know ... 20 days on one sensor is unheard of for me, but this one has been good and I've rotated out to the back of my arms.

**Comment by efidoman** [comment]
Waiting to see now ...
08/19 20:08:02 Using Auto Sensor Restart: : true

**Comment by efidoman** [comment]
Here's where it sent all the messages stop/start/calibrate/calibrate

[{"date":"1566252761964","type":"StopSensor"}]
[{"date":"1566252763056","type":"StartSensor","sensorSerialCode":""}]
calibration treatment posted to /root/myopenaps/monitor/xdripjs/calibration.json - record is below
[{"date":1566259964000,"type":"CalibrateSensor","glucose":142}]
calibration treatment posted to /root/myopenaps/monitor/xdripjs/calibration.json - record is below
[{"date":1566259965000,"type":"CalibrateSensor","glucose":142},{"date":1566259964000,"type":"CalibrateSensor","glucose":142}]

here's the next loop where it did them ...

messages passed to logger: [{"date":"1566252761964","type":"StopSensor"},{"date":"1566252763056","type":"StartSensor","sensorSerialCode":""},{"date":1566259965000,"type":"CalibrateSensor","glucose":142},{"date":1566259964000,"type":"CalibrateSensor","glucose":142}]

and here's where it transmitted them
2019-08-20T00:17:40.797Z bluetooth-manager Rx 290001cd952c00ffffffff18b32c00a2fd
2019-08-20T00:17:40.801Z transmitter got message:  SessionStopRxMessage {
  status: 0,
  received: 1,
  requestedStopTime: 2921933,
  sessionStopTime: 4294967295,
  transmitterTime: 2929432 }
logger message received: {"time":"1566252761964"}
2019-08-20T00:17:40.806Z transmitter Sending command: StartSensor
2019-08-20T00:17:40.812Z bluetooth-manager Tx 26ce952c00243c5b5d0000dcc6
2019-08-20T00:17:40.857Z bluetooth-manager Rx 270001ce952c00ce952c0018b32c00b378
2019-08-20T00:17:40.859Z transmitter got message:  SessionStartRxMessage {
  status: 0,
  received: 1,
  requestedStartTime: 2921934,
  sessionStartTime: 2921934,
  transmitterTime: 2929432 }
logger message received: {"time":"1566252763056"}
2019-08-20T00:17:40.861Z transmitter Sending command: CalibrateSensor
2019-08-20T00:17:40.865Z bluetooth-manager Tx 348e00f0b12c00c1d9
2019-08-20T00:17:40.887Z bluetooth-manager Rx 3500097cbf
2019-08-20T00:17:40.889Z transmitter got message:  CalibrateGlucoseRxMessage {}
logger message received: {"time":1566259965000}
2019-08-20T00:17:40.890Z transmitter Sending command: CalibrateSensor
2019-08-20T00:17:40.892Z bluetooth-manager Tx 348e00efb12c008816
2019-08-20T00:17:40.917Z bluetooth-manager Rx 350010643c
2019-08-20T00:17:40.918Z transmitter got message:  CalibrateGlucoseRxMessage {}
logger message received: {"time":1566259964000}
2019-08-20T00:17:40.923Z transmitter no backfill needed

Same Logger Loop - it went into first calibration ...

08/19 20:17:41 Sensor state = First calibration

Next loop it went directly to second calibration

08/19 20:22:42 Sensor state = Second calibration

Well, it mostly worked this time. Maybe 1 second isn't enough time between first and second calibration?

The sensor seemed to accept both calibrations. Ideas?

**Comment by tynbendad** [comment]
Yeah, just try spacing them out a bit, even the first one may have been skipped if the start took a bit to take affect.

**Comment by efidoman** [comment]
Perhaps I can do 5 seconds between the stop, start, and first calibration ... I do believe that they worked with 1 second between. And then do 61 seconds between the first and second calibrations. Let me try that now. This sensor is still going strong so it is a good one to use for these tests.

**Comment by tynbendad** [comment]
Isn't the tx just awake for 20-30 seconds?

**Comment by efidoman** [comment]
Same results exactly ... everything worked, the age of the sensor is showing 2 hours (so stop/start worked) and the first calibration took but the 2nd didn't. I'll test again later.

**Comment by efidoman** [comment]
@tynbendad - how do you suggest we handle the case where filtered/unfiltered are null but we have a valid glucose when it comes to noise calculation? Just set noise to unknown?

**Comment by tynbendad** [comment]
Dexcom suppresses glucose when noisy so we could either life with just that, as if openaps was getting data from receiver/share server, or do our own noise calcs based on glucose instead of unfiltered. I vote for just like a share server fwiw.

**Comment by efidoman** [comment]
I think the most appropriate setting for noise in this case is "Clean" because the tx will not give a valid glucose if not "Clean".

**Pull request merged**